src/encoded/types/cart.py

```diff
@@ -48,51 +54,94 @@ class Collection(Item.Collection):
         pass
 
 
-def _get_carts_by_user(request, userid):
-    request.datastore = 'database'
-    return [
-        request.resource_path(v, '')
-        for k, v in request.registry[COLLECTIONS]['cart'].items()
-        if v.properties['submitted_by'] == userid
-    ]
-
-
-def _create_cart(request, user):
+def _create_cart(request, user, name=None, identifier=None, status=None):
     carts = request.registry[COLLECTIONS]['cart']
     user_props = request.embed(request.resource_path(user), '@@object')
+    cart_name = name or '{} cart'.format(user_props['title'])
+    initial_cart = {
+        'submitted_by': str(user.uuid),
+        'status': status or 'current',
+        'name': cart_name,
+        'elements': []
+    }
+    if identifier:
+        initial_cart['identifier'] = identifier
     cart = create_item(
         carts.type_info,
         request,
-        {
-            'submitted_by': str(user.uuid),
-            'status': 'current',
-            'name': '{} cart'.format(user_props['title']),
-            'elements': []
-        }
+        initial_cart,
     )
-    return request.resource_path(cart)
+    cart_path = request.resource_path(cart)
+    return cart_path
 
 
-@view_config(context=Cart.Collection, request_method='GET', permission='save-carts', name='get-cart')
-def get_or_create_cart_by_user(context, request):
+def _get_userid(request):
     userid = [
         p.replace('userid.', '')
         for p in request.effective_principals
         if p.startswith('userid.')
     ]
     if not userid:
         raise HTTPBadRequest()
-    else:
-        userid = userid[0]
+    return userid[0]
+
+
+def _get_user(request, userid):
     user = request.registry[COLLECTIONS]['user'].get(userid)
     if not user:
         raise HTTPBadRequest()
-    carts = _get_carts_by_user(request, userid)
+    return user
+
+
+@view_config(context=Cart.Collection, request_method='GET', permission='save-carts', name='get-cart')
+def get_or_create_cart_by_user(context, request):
+    userid = _get_userid(request)
+    user = _get_user(request, userid)
+    carts = get_cart_objects_by_user(request, userid, ['disabled', 'deleted '])
     if not carts:
         cart = _create_cart(request, user)
+        cart_atids = None
+    else:
+        cart_atids = [cart['@id'] for cart in carts]
+    request.response.status = 200
+    return {
+        'status': 'success',
+        '@type': ['result'],
+        '@graph': cart_atids or [cart]
+    }
+
+
+@view_config(context=Cart.Collection, request_method='PUT', permission='save-carts', name='put-cart')
+def create_cart_by_user(context, request):
+    userid = _get_userid(request)
+    user = _get_user(request, userid)
+    user_carts = get_cart_objects_by_user(request, userid, ['deleted'] if not 'group.admin' in request.effective_principals else [])
+    cart_atids = [cart['@id'] for cart in user_carts]
+    cart_status = request.json.get('status', None)
```

Comment:
Get returns None by default so no need to specify for most of these.

Reply:
I couldn’t find confirmation this is true, but oh well.

Reply:
Pyramid uses request object from Webob that sets `.json` and `.json_body` to the same thing: https://github.com/Pylons/webob/blob/b5297529ef5a8dd3dd84c6cfa41afd101acfa8a2/src/webob/request.py#L739

This just uses the Python JSON library that handles input as follows:
```
JSON	Python
object	dict
array	list
string	unicode
number (int)	int, long
number (real)	float
true	True
false	False
null	None
```

So you should be getting a dict back based on what you have PUT to the endpoint. Python dicts have the get method which you are using, defined as follows:

> get(key[, default])
> Return the value for key if key is in the dictionary, else default. If default is not given, it defaults to None, so that this method never raises a KeyError.

https://docs.python.org/3/library/stdtypes.html#dict.get

Reply:
Ah thanks, that helps. I thought that .get was something specific to request.json, not a general dict thing.